### ISSUE TITLE
Pyright fails to identify return type mismatch

### ISSUE THREAD

**Issue by ianliu** [microsoft/pyright#3899]
**Describe the bug**

The `Result.map` function shouldn't type-check. It is returning `Result[T, E]` on the first branch of the if statement, whereas it should be returning `Result[U, E]`:

```python
from __future__ import annotations
from dataclasses import dataclass
from collections.abc import Callable
from typing import Generic, TypeVar

T = TypeVar("T")
U = TypeVar("U")
E = TypeVar("E")

@dataclass
class Ok(Generic[T]):
    value: T

@dataclass
class Err(Generic[E]):
    value: E

@dataclass
class Result(Generic[T, E]):
    value: Ok[T] | Err[E]

    def map(self, fn: Callable[[T], U]) -> Result[U, E]:
        print(fn)
        if isinstance(self.value, Ok):
            return Result(self.value)
        else:
            return Result(self.value)

```

**Expected behavior**
Pyright should warn about a type mismatch at the return of the map function.

**VS Code extension or command-line**
Running pyright from the command line, version 1.1.269 from Archlinux.

**Additional context**
Mypy works as expected, as seen here: https://mypy-play.net/?mypy=latest&python=3.10&gist=aa64cc574a330249ae420d68fda5b4e2

**Comment by ianliu** [comment]
The following also has the same behavior:

```python
    ...
    def map(self, _: Callable[[T], U]) -> Result[U, E]:
        return Result(self.value)
```

But this works as expected:

```python
    ...
    def map(self, _: Callable[[T], U]) -> Result[U, E]:
        return self
```

> Expression of type "Self@Result[T@Result, E@Result]" cannot be assigned to return type "Result[U@map, E@Result]"
>      TypeVar "T@Result" is invariant
>        Type "T@Result" cannot be assigned to type "U@map"

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.270, which I just published. It will also be included in a future release of pylance.
